[RNMobile] Embed block - Link in Block Settings

jd-alexander opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->
`Gutenberg-Mobile` PR https://github.com/wordpress-mobile/gutenberg-mobile/pull/4189
## Description
This PR is ready for a first pass while some minor issues are investigated 🙇🏾 

<!-- Please describe what you have changed or added -->
This PR introduces the Link in the Block Settings of each Embed block. Since the Link is now within the Block Settings component, we no longer need to display the Pencil icon ( Edit action) with the Toolbar, therefore this functionality was removed. To facilitate these changes, the `embed-controls.js` component that was shared between the web and mobile, was duplicated and modifications were made to the `embed-controls.native.js` variant to support the Block Settings and Toolbar behavior specific to mobile as described in the associated GB-Mobile ticket. 

## How has this been tested?

### Add Embed block - bottom sheet functionality remains ✅

The main purpose of this test case is to ensure that the changes made here did not impact the existing functionality. 

1. Add an Embed block. 
2. The bottom sheet should auto-open and be focused. 
3. Add a link to the bottom sheet and dismiss it.
4. Notice the Embed placeholder transforming to an Embed Preview. 

https://user-images.githubusercontent.com/1509205/139797669-bc71766d-6fe6-4288-a1bb-2f99a5f43470.mov

### Edit Embed block - change the link in Block Settings ✅

1. Using the steps above, create an Embed block and populate it with a preview. 
2. Click on the Gear/Settings icon. 
3. Notice the Link Settings within the Block Settings. 
4. Change the link and dismiss the bottom sheet. 
5. The Embed Preview content should be updated.
6. Notice that the **Pencil icon** does not **exist**. 

https://user-images.githubusercontent.com/14905380/140941412-3d50a725-4318-4648-896b-96752e925b66.mp4

### Edit Embed block - Invalid format link does not lead to validation error 🟢 

1. Using the steps above, create an Embed block and populate it with a preview. 
2. Click on the Gear/Settings icon. 
3. Notice the Link Settings within the Block Settings. 
4. Change the link to an invalid one and dismiss the bottom sheet (example: `http://`). 
5. The Embed Preview tries to load the content.
6. Notice that the URL error notice is displayed. 
7. ~~Since the Pencil icon is no longer present we are now left in a state where the link cannot be edited.~~

~~I think this occurs because the `LinkPicker` adds `http://` to all entries hence the `embed-bottomsheet` logic believes the URL is true. I will investigate this further.
https://github.com/WordPress/gutenberg/blob/3e371ca996c8cf78e09f0eeb4ff9051af3a0d5a7/packages/block-library/src/embed/embed-bottom-sheet.native.js#L51~~

https://user-images.githubusercontent.com/14905380/140942059-dad9ba4f-1b69-470d-8a4b-3d513660d77d.mp4

### Edit Embed block - Invalid link  🟢 

1. Using the steps above, create an Embed block and populate it with a preview. 
2. Click on the Gear/Settings icon. 
3. Notice the Link Settings within the Block Settings. 
4. Change the link to an invalid one and dismiss the bottom sheet (example: `https://example.com`). 
5. The Embed Preview tries to load the content.
6. Notice the block shows an error message about being unable to embed media.
7. Click on "More options".
8. Click on the "Edit link" option.
9. Set a valid URL and dismiss the bottom sheet.
10. The Embed Preview content should be updated.

https://user-images.githubusercontent.com/14905380/140943215-4c85353d-fc00-46cc-b7dc-93999736a8fa.mp4

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->


## Types of changes
1. A native `EmbedControl` was created. 
2. The Pencil - Edit icon was removed. 
3. A Link Picker is now used within the Block Settings to change the URL. 

<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jd-alexander committed:
Created a native version of embed-controls

jd-alexander committed:
added withBottomSheet flag so embed link works in both modes.

jd-alexander committed:
integrate EmbedBottomSheet with EmbedControls in Embed functionality

jd-alexander committed:
Removed the edit button from the toolbar. Removed all related code.

jd-alexander committed:
onClose calls onDismiss to perform necessary url validation & submission

jd-alexander committed:
the responsive condition should encapsulate the media settings only.

jd-alexander committed:
utilize the LinkPicker when the Embed URL is being edited.

jd-alexander committed:
Removed unneeded className in the native embed-controls.

jd-alexander committed:
Added entry to the CHANGELOG.

jd-alexander commented on packages/block-library/src/embed/edit.native.js:
```diff
@@ -236,10 +236,6 @@ const EmbedEdit = ( props ) => {
 		<>
 			{ showEmbedPlaceholder ? (
 				<>
-					<EmbedControls
```
Even though the `Edit Button` functionality is being removed, I still kept the `setIsEditingURL` behavior. I will utilize the PR associated with these changes to verify that it is still necessary and functions as expected.

jd-alexander commented on packages/block-library/src/embed/edit.native.js:
```diff
@@ -262,12 +258,23 @@ const EmbedEdit = ( props ) => {
 						allowResponsive={ allowResponsive }
 						toggleResponsive={ toggleResponsive }
-						switchBackToURLInput={ () => setIsEditingURL( true ) }
+						bottomSheetLabel={ bottomSheetLabel }
+						url={ url }
+						onEmbedBottomSheetSubmit={ ( value ) => {
+							// The order of the following calls is important, we need to update the URL attribute before changing `isEditingURL`,
+							// otherwise the side-effect that potentially replaces the block when updating the local state won't use the new URL
+							// for creating the new block.
+							setAttributes( { url: value } );
+							setIsEditingURL( false );
+						} }
+						onEmbedBottomSheetClose={ () =>
+							setShowEmbedBottomSheet( false )
+						}
+						showEmbedBottomSheet={ showEmbedBottomSheet }
```
These props are similar to how the `EmbedBottomSheet` is used within this Component. Let me know if the logic sharing approach here is fine or if there's a more optimal way to do this.

fluiddot replied:
Now that the visibility of `EmbedBottomSheet` component is controlled by `EmbedControls`, probably we could render `EmbedBottomSheet` within the `EmbedControls` component and only expose here a prop to notify when the URL is edited, wdyt?

jd-alexander replied:
@fluiddot I think that approach is fine. Do we want to link picking behavior to be the same for both adding and editing a link? I assume that would make the experience more optimal. However, I noticed that when adding a link the embed bottom sheet would accept the input directly but when editing a link as done here, there would be a picker for that link because that is the approach that I see being utilized in other Block Settings. So the question is do we utilize inline editing instead of a Picker for both experiences? I think doing this would allow the experience to be optimal. ( At first, I had tried this but I think how the attributes were being set was possibly causing an issue so I didn't pursue it)

Now when I think about it, I agree with the case where if we utilize a prop to notify when a URL is being edited then, only at that time would we need to render the link editing component within the `BlockSettings` `PanelBody` otherwise we return the bottom sheet. i.e a function that does something similar to the link settings logic below. 

https://github.com/WordPress/gutenberg/blob/fbac29b09913c60f3004eb979542f851daf87f7a/packages/components/src/mobile/link-settings/index.native.js#L285

fluiddot replied:
> @fluiddot I think that approach is fine. Do we want to link picking behavior to be the same for both adding and editing a link? I assume that would make the experience more optimal. However, I noticed that when adding a link the embed bottom sheet would accept the input directly but when editing a link as done here, there would be a picker for that link because that is the approach that I see being utilized in other Block Settings. So the question is do we utilize inline editing instead of a Picker for both experiences? I think doing this would allow the experience to be optimal. ( At first, I had tried this but I think how the attributes were being set was possibly causing an issue so I didn't pursue it)

I think it makes sense to provide the same experience in both cases. I'd stick with the option where the user can edit directly the URL (as we're currently providing). The one that presents the picker can be confusing because it's related to linking the block to an URL, but in this case, we're editing the source of the embed.

> Now when I think about it, I agree with the case where if we utilize a prop to notify when a URL is being edited then, only at that time would we need to render the link editing component within the BlockSettings PanelBody otherwise we return the bottom sheet. i.e a function that does something similar to the link settings logic below.

Yep, exactly. I hope this doesn't introduce undesired side effects but looks like a promising way to go 👍 . Let me know if I can help anyhow in the implementation 🙇 .

jd-alexander replied:
> I think it makes sense to provide the same experience in both cases. I'd stick with the option where the user can edit directly the URL (as we're currently providing). The one that presents the picker can be confusing because it's related to linking the block to an URL, but in this case, we're editing the source of the embed.

Agreed! 

> Yep, exactly. I hope this doesn't introduce undesired side effects but looks like a promising way to go 👍 . Let me know if I can help anyhow in the implementation 🙇 .

Thanks for confirming 🙇🏾 My goal was to take a look at it after I am done WPiOS testing efforts and crash monitoring. If you would want to do the refactor and it doesn't take much effort from your side I am more than happy to review the changes afterward in an effort to keep things moving.

fluiddot replied:
> Thanks for confirming 🙇🏾 My goal was to take a look at it after I am done WPiOS testing efforts and crash monitoring. If you would want to do the refactor and it doesn't take much effort from your side I am more than happy to review the changes afterward in an effort to keep things moving.

I'd like first to wrap up first the tests tasks but after that, I'll be more than happy to help with this implementation. I'll let you know if I start tackling it 👍 .

jd-alexander commented on packages/block-library/src/embed/embed-bottom-sheet.native.js:
```diff
@@ -56,6 +63,17 @@ const EmbedBottomSheet = ( { value, label, isVisible, onClose, onSubmit } ) => {
+	 * If the Embed Bottom Sheet component does not utilize a bottom sheet then the onDismiss action is not
+	 * called. Here we are wiring the onDismiss to the onClose callback that gets triggered when input is submitted.
+	 */
+	const performOnCloseOperations = () => {
```
The idea here is that when the Link Picker is being used the `onDimiss` is not called. Therefore when `onClose` is called, it functions as `onDismiss` as well.

fluiddot commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -10,7 +10,7 @@ For each user feature we should also add a importance categorization label  to i
 
 ## Unreleased
-
+[*] [Embed block] Included Link in Block Settings [#36099]
```
I think we're missing a new line here to separate the `Unreleased` from the `1.65.0`.

fluiddot reviewed (commented):
Thanks @jd-alexander for tackling this ❤️!

I've done a quick review and I'd like to propose a different approach where the `EmbedBottomSheet` would be rendered within the `EmbedControls`. I think it would be possible but I might be overlooking a potential blocker, let me know what you think, I'll be more than happy to help to land it 🙇 .

fluiddot commented:
@jd-alexander regarding the issue about the invalid links not leading to a validation error, what URLs did you use for testing? I recall I used the value `http://` for this case as referenced in [this PR](https://github.com/WordPress/gutenberg/pull/33253) because, as you commented, the `LinkPicker` component automatically adds the `https://` prefix if not provided.

jd-alexander commented:
> @jd-alexander regarding the issue about the invalid links not leading to a validation error, what URLs did you use for testing? I recall I used the value `http://` for this case as referenced in [this PR](https://github.com/WordPress/gutenberg/pull/33253) because, as you commented, the `LinkPicker` component automatically adds the `https://` prefix if not provided.

For this particular case, I didn't use a URL, I had just typed in a random set of letters. However, this error only occurs when the `hasPicker` prop is passed. Since we are using the inline editing experience as confirmed in [this comment](https://github.com/WordPress/gutenberg/pull/36099#discussion_r742112262) then we won't encounter this issue as our existing implementation with the `LinkSettingsNavigation` within the `EmbedBottomSheet` works as expected.

fluiddot commented:
> For this particular case, I didn't use a URL, I had just typed in a random set of letters. However, this error only occurs when the hasPicker prop is passed. Since we are using the inline editing experience as confirmed in this comment then we won't encounter this issue as our existing implementation with the LinkSettingsNavigation within the EmbedBottomSheet works as expected.

Ok, I'm glad to know that it's not happening in the inline editing experience. In any case, it would be great to double-check this once we finish the implementation, thanks for clarifying it 🙇 .

fluiddot commented:
@jd-alexander I'm going to devote some time today to implement the approach we discussed in [this comment](https://github.com/WordPress/gutenberg/pull/36099#discussion_r742127821).

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-link-to

# Conflicts:
#	packages/block-library/src/embed/embed-bottom-sheet.native.js
#	packages/react-native-editor/CHANGELOG.md

fluiddot committed:
Refactor EmbedBottomSheet component

The EmbedBottomSheet component has been also renamed to EmbedLinkSettings because the component is no longer strictly tied to the bottom sheet component.

fluiddot commented on packages/block-library/src/embed/edit.native.js:
```diff
@@ -196,6 +196,14 @@ const EmbedEdit = ( props ) => {
 		isEditingURL,
 	] );
 
+	const onEditURL = useCallback( ( value ) => {
+		// The order of the following calls is important, we need to update the URL attribute before changing `isEditingURL`,
+		// otherwise the side-effect that potentially replaces the block when updating the local state won't use the new URL
+		// for creating the new block.
+		setAttributes( { url: value } );
+		setIsEditingURL( false );
+	}, [] );
```
I've extracted the edit URL handler as now it's used in two places, the embed controls and the bottom sheet.

fluiddot commented on packages/block-library/src/embed/embed-controls.native.js:
```diff
@@ -0,0 +1,65 @@
+						value={ url }
+						label={ linkLabel }
+						onSubmit={ ( value ) => {
+							closeSettingsBottomSheet();
```
I had to force closing the bottom sheet when the URL changes because in some cases, this implies that the block will be recreated and new preview data fetched hence, the bottom sheet will be empty during the transition.

fluiddot commented on packages/block-library/src/embed/embed-link-settings.native.js:
```diff
@@ -0,0 +1,95 @@
+import { store as noticesStore } from '@wordpress/notices';
+import { useCallback, useRef, useState } from '@wordpress/element';
+
+const EmbedLinkSettings = ( {
```
This component is a copy of the `EmbedBottomSheet` component but with the following tweaks:

### URL is now kept in two places
The URL is now kept as an instance variable via the `useRef` and in the component state. The reason for this is that when `onDismiss` is called, I realized that the previous `url` state value ([reference](https://github.com/WordPress/gutenberg/pull/36099/files?authenticity_token=Ah3%2FJt4s4h5hECy91JtulYIvCsOA1prp%2BgPsIMxjD23WBg3FzYJhCbA8f0qyAKOE%2BlHAyPqav9wNMjPuwFlWIQ%3D%3D&file-filters%5B%5D=.js&file-filters%5B%5D=.md#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fL15)) isn't updated even after the `setURL` call ([reference](https://github.com/WordPress/gutenberg/pull/36099/files?authenticity_token=Ah3%2FJt4s4h5hECy91JtulYIvCsOA1prp%2BgPsIMxjD23WBg3FzYJhCbA8f0qyAKOE%2BlHAyPqav9wNMjPuwFlWIQ%3D%3D&file-filters%5B%5D=.js&file-filters%5B%5D=.md#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fL60)). Not sure why is this happening but I decided to keep the internal value of the URL as an instance value, which is used in the different handlers, and have a URL in the component state only for passing it to the `LinkSettingsNavigation` component.

### Autofocus prop
The link field shouldn't be autofocused when editing an URL via the block settings, so I added a prop to control it.

### onDismiss handler simplified
I noticed that the code could be simplified as there was an `else` case that was redundant ([reference](https://github.com/WordPress/gutenberg/pull/36099/files?authenticity_token=Ah3%2FJt4s4h5hECy91JtulYIvCsOA1prp%2BgPsIMxjD23WBg3FzYJhCbA8f0qyAKOE%2BlHAyPqav9wNMjPuwFlWIQ%3D%3D&file-filters%5B%5D=.js&file-filters%5B%5D=.md#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fL53-L56)).

fluiddot commented on packages/block-library/src/embed/embed-link-settings.native.js:
```diff
@@ -0,0 +1,95 @@
+		onSubmit( url.current );
+	}, [ onSubmit, value ] );
+
+	/**
+	 * If the Embed Bottom Sheet component does not utilize a bottom sheet then the onDismiss action is not
+	 * called. Here we are wiring the onDismiss to the onClose callback that gets triggered when input is submitted.
+	 */
+	const performOnCloseOperations = useCallback( () => {
+		if ( onClose ) {
+			onClose();
+		}
+
+		if ( ! withBottomSheet ) {
+			onDismiss();
+		}
+	}, [ onClose ] );
```
@jd-alexander I kept the code you introduced as the `onDismiss` is not triggered when displaying the link field in the block settings.

fluiddot commented:
@jd-alexander I've made a refactor in the `EmbedBottomSheet` component (check previous comments), however, I noticed the following issues that should be addressed:

### No way to edit URL if the preview fails ✅ 
After removing the edit URL button, when using a bad URL (example: `https://example.com`) there's no option to edit the URL. I see two options to solve it:
1. Revert the edit URL button removal.
2. Add an Edit URL option to the "More options" picker.

<details><summary>Click to see screenshot</summary>

<img src=https://user-images.githubusercontent.com/14905380/140401895-48db2e8d-3ddc-40d2-9c0f-f9d1b3dfa46c.png width=400>

</details>

**EDIT:** Addressed in commit: https://github.com/WordPress/gutenberg/commit/87375985de697617fdf1c3b93fa30a568f0800cf

### Previous URL is auto-populated after setting an empty URL ✅ 
I couldn't find the culprit of this bug yet, here are the reproduction steps:
1. Add an embed block.
2. Set a URL (example: `https://www.youtube.com/watch?v=ssfHW5lwFZg`).
3. Open the block settings via the gear icon button.
4. Remove the link and dismiss the bottom sheet.
5. Observe that the embed block is now empty.
6. Tap on "ADD LINK".
7. **Observe that the previous URL is already set in the link field when it should be empty.**

**EDIT:** Addressed in commit: https://github.com/WordPress/gutenberg/commit/87f14089b4f2bd8ebbafcd0834f16a3a26623361

### Some tests of the embed block are failing
We would need to update the integration tests of the embed block with the changes introduced in this PR.

jd-alexander commented:
> No way to edit URL if the preview fails
After removing the edit URL button, when using a bad URL (example: https://example.com) there's no option to edit the URL. I see two options to solve it:
> 1. Revert the edit URL button removal.
> 2. Add an Edit URL option to the "More options" picker.

Indeed. I would utilize option two here. It's easier to spot from a discoverability standpoint and all other scenarios will allow the link to be edited via the Block Settings flow. 

> Previous URL is auto-populated after setting an empty URL

I will reproduce this and have a look. 

> We would need to update the integration tests of the embed block with the changes introduced in this PR.

Thanks for noting 🙇🏾

jd-alexander committed:
Added Edit Link functionality to Error Picker options.

jd-alexander commented:
@fluiddot "No way to edit URL if the preview fails" was resolved in 8737598
### Utilize Edit Link within the More options Picker to edit URL if the preview fails

1. Add an Embed block. 
2. Add an invalid URL. 
3. Notice the preview fails. 
4. Select more options so that the Picker is presented. 
5. Select Edit link. 
6. Edit the link in the `EmbedLinkSettings` bottom sheet with a valid URL. 
7. Notice that the Preview is rendered as expected.


https://user-images.githubusercontent.com/1509205/140423028-d598fe70-c224-4599-9cd8-00dfab579eb2.mov

jd-alexander commented:
> Previous URL is auto-populated after setting an empty URL

I investigated this and I am wondering if the value is being retained here

```
	const url = useRef( value );
	const [ inputURL, setInputURL ] = useState( value );
```
even after they are being cleared in `onSetAttributes` due to some inconsistency with how the state updates are being done.

fluiddot committed:
Update url when value changes

fluiddot commented:
> > Previous URL is auto-populated after setting an empty URL
> 
> I investigated this and I am wondering if the value is being retained here
> 
> ```
> 	const url = useRef( value );
> 	const [ inputURL, setInputURL ] = useState( value );
> ```
> 
> even after they are being cleared in `onSetAttributes` due to some inconsistency with how the state updates are being done.

I realized that now, since we're rendering two `EmbedLinkSettings` components, we need to synchronize the URL value on both instances to prevent the issue. For this reason, I added a side effect in https://github.com/WordPress/gutenberg/pull/36099/commits/87f14089b4f2bd8ebbafcd0834f16a3a26623361 to keep the value updated when the URL changes, so the issue is addressed ✅ .

jd-alexander committed:
Utilized the Block Settings Bottom Sheet for tests that are editing URL

jd-alexander commented:
@fluiddot I started the effort of updating the failing tests acd9259. 


1. I am faced with an issue I have been investigating so far. 

This issue below is occurring when I try to utilize `getByA11yLabel( Embed link, ${ initialURL })` to start editing the URL.
```
Expected 1 but found 2 instances with accessibilityLabel "Embed link, https://youtu.be/BAD_URL"
```

I utilized `debug` from the `render` to see the tree of the `Views` and I am seeing that two View instances have the label. I know that the primary view in question here is the `TextControl` 

https://github.com/WordPress/gutenberg/blob/a9b7e98ecf48ae134e42074abb23115fd31c9c7a/packages/components/src/mobile/link-settings/index.native.js#L233

I see that both the `TextControl` and the underlying `Cell` are using the label. I am wondering if that may be related. 
I will take another look tomorrow but if you are able to have a look, that would be helpful. Thanks. 

2. Within the tests, I am following the previous test behavior of waiting for the bottom sheet to become visible and also dismissing it. However, I am seeing where they are other tests in the codebase that do not close the bottom sheet after opening. Is it fine to keep this functionality within the tests that are utilizing the Block Settings?

jd-alexander committed:
Removed unneeded whiteline

jd-alexander committed:
Merge branch 'trunk' into rnmobile/embed-block-link-to

fluiddot committed:
Fix indentation and lint issues

fluiddot committed:
Use within to query edit link field

fluiddot committed:
Update responsive not supported test case

fluiddot committed:
Expect for link field instead block settings button

This change prevents potential act warnings after setting a URL in an empty embed block.

fluiddot committed:
Invalidate core-data store before each test

fluiddot commented:
@jd-alexander Thank you very much for starting addressing the issues on the integration tests 🥇!

> 1. I am faced with an issue I have been investigating so far.
> 
> This issue below is occurring when I try to utilize `getByA11yLabel( Embed link, ${ initialURL })` to start editing the URL.
> 
> ```
> Expected 1 but found 2 instances with accessibilityLabel "Embed link, https://youtu.be/BAD_URL"
> ```
> 
> I utilized `debug` from the `render` to see the tree of the `Views` and I am seeing that two View instances have the label. I know that the primary view in question here is the `TextControl`
> 
> https://github.com/WordPress/gutenberg/blob/a9b7e98ecf48ae134e42074abb23115fd31c9c7a/packages/components/src/mobile/link-settings/index.native.js#L233
> 
> I see that both the `TextControl` and the underlying `Cell` are using the label. I am wondering if that may be related. I will take another look tomorrow but if you are able to have a look, that would be helpful. Thanks.

Right, there are two instances because the component `EmbedLinkSettings` is rendered in two places, the edit component ([reference](https://github.com/WordPress/gutenberg/blob/6fbd7ebe9ede9653cda1a6a6cd59642edb2a7d75/packages/block-library/src/embed/edit.native.js#L300-L309)) to edit the URL upon block creation or when it's empty, and within the embed controls to edit the URL via the block settings ([reference](https://github.com/WordPress/gutenberg/blob/6fbd7ebe9ede9653cda1a6a6cd59642edb2a7d75/packages/block-library/src/embed/embed-controls.native.js#L51-L58)).
 
In this case, we can solve the issue by using the [`within` function](https://callstack.github.io/react-native-testing-library/docs/api#within-getqueriesforelement) from the React Native testing library, which allows us to perform a query within an element. I went ahead and applied this change in https://github.com/WordPress/gutenberg/pull/36099/commits/e27ff9a56088ec92ae5b531612e050ed1b32c1d4.

> 2. Within the tests, I am following the previous test behavior of waiting for the bottom sheet to become visible and also dismissing it. However, I am seeing where they are other tests in the codebase that do not close the bottom sheet after opening. Is it fine to keep this functionality within the tests that are utilizing the Block Settings?

In general, it shouldn't be necessary to wait for visibility changes in the bottom sheet, however in this case, as we use the dismiss action to perform editing the URL we need it.

<hr>

Apart from introducing the use of `within` function for addressing the first issue, I also pushed the following changes to solve other issues I found out debugging further the tests:
1. [Update responsive not supported test case](https://github.com/WordPress/gutenberg/pull/36099/commits/3023e204cfc785fa94d5050fb26b3cbc32f4cf65): With the changes in this PR, the block settings button is now always enabled, so this test case now checks if the `Media settings` panel is present depending on the responsive support.
2. Expect for link field instead block settings button (https://github.com/WordPress/gutenberg/pull/36099/commits/4ae85dcadb9605373093d804665e43f45efd298a, https://github.com/WordPress/gutenberg/pull/36099/commits/cc80690a2a4640ffdf6625a3dedf6e8ea1f86fa1): I noticed some `act` warnings while testing produced by the `set URL upon block insertion` test cases, which I managed to fix by using a different expectation.
3. [Invalidate core-data store before each test](https://github.com/WordPress/gutenberg/pull/36099/commits/347f7ff3297ec7aa263768be651daf0e6dbc05b7): By only enabling some of the test cases, I realized that it was required to invalidate the entire `core-data` store to prevent caching the network requests made by the embed block. We were already invalidating the requests for fetching the embed previews but not the ones for the theme support. Here is the list of the test cases I disabled for testing this:
    1. `displays cannot embed on the placeholder if preview data is null`
    2. `toggles resize for smaller devices media settings` 

@jd-alexander I think the PR is finally ready for a final review, let me know if you could take a look, thanks 🙇 .

fluiddot committed:
Expect for link field instead block settings button in more tests

fluiddot committed:
Add comments to test actions

fluiddot commented:
I noticed that the Performance and End-to-End Tests are also failing in `trunk`, so failures aren't caused by the changes in this PR. Once the PR is ready to merge, we should review if the issues are addressed and update the PR with `trunk`, if not, we could consider skipping the failed tests and proceed with the merge.

fluiddot committed:
Add set empty URL test case

fluiddot committed:
Add edit URL when request fails test case

fluiddot commented:
@jd-alexander I updated the test cases described in the PR description due to the recent changes and added one for verifying that it allows editing the link after setting an invalid link.

Additionally, I added a couple of integration tests:

- [Add set empty URL test case](https://github.com/WordPress/gutenberg/pull/36099/commits/69c9dbd3a932a4e2aa6e3fc3ec73b971f3f80143)
- [Add edit URL when request fails test case](https://github.com/WordPress/gutenberg/pull/36099/commits/aa754c9c759fd4269c671187e089dc0e7654624c)

fluiddot commented on packages/block-editor/src/components/block-settings/container.native.js:
```diff
@@ -43,6 +43,7 @@ function BottomSheetSettings( {
 			hideHeader
 			contentStyle={ styles.content }
 			hasNavigation
+			testID="block-settings"
```
I'm wondering if this test ID should be more unique by renaming it to `block-settings-modal`, I'm concerned that it might introduce conflicts with other test IDs in the future, wdyt?

jd-alexander replied:
That is a good idea. I will make this change shortly.

jd-alexander replied:
I made the suggested change and re-ran the tests with `npm run native test packages/block-library/src/embed/test/index.native.js` and all the tests passed. 👍🏾

fluiddot reviewed (commented):
I run all test cases and everything is working nice 🎊 , I only found a minor issue related to displaying the invalid URL error notice twice, here are the reproduction steps:
1. Add an embed block.
2. Set a valid URL.
3. Open block settings.
4. Change embed link to an invalid format URL (example: `http://`).
5. Tap on the enter key on the keyboard instead of dismissing the bottom sheet.
6. Observe that the `Invalid URL. Please enter a valid URL.` error message is displayed twice.

https://user-images.githubusercontent.com/14905380/140948001-bc8eeee9-16dd-4506-afb8-0d57f11bb39d.mp4

After debugging this issue, I noticed that the `performOnCloseOperations` handler is being called twice when editing the URL by tapping on the enter key. However, when dismissing the bottom sheet is only called once 🙃 .

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-link-to

jd-alexander committed:
Created a ref to store if the onCloseSettingsSheet is consumed.

jd-alexander commented:
> I run all test cases and everything is working nice 🎊 , I only found a minor issue related to displaying the invalid URL error notice twice

Thanks for spotting this error @fluiddot I resolved this issue by storing if the bottom sheet's close event was consumed in an instance variable via `useRef`. 3255826

https://github.com/WordPress/gutenberg/blob/32558261ca127553258b3df7b22983d3f331654f/packages/components/src/mobile/link-settings/index.native.js#L96

The reason this occurs is because `onClose` is triggered in these two locations within the `LinkSettings` component. 

1. https://github.com/WordPress/gutenberg/blob/32558261ca127553258b3df7b22983d3f331654f/packages/components/src/mobile/link-settings/index.native.js#L102

2. https://github.com/WordPress/gutenberg/blob/ab7e8aa0312b2e5264af57c771a04cd9702370d3/packages/components/src/mobile/link-settings/index.native.js#L239

Let me know if this approach is optimal. Thank you 🙇🏾

jd-alexander committed:
Added "modal" to  "block-settings" TestId to make "block-settings-modal"

jd-alexander commented:
> @jd-alexander I updated the test cases described in the PR description due to the recent changes and added one for verifying that it allows editing the link after setting an invalid link.
> 
> Additionally, I added a couple of integration tests:
> 
> * [Add set empty URL test case](https://github.com/WordPress/gutenberg/pull/36099/commits/69c9dbd3a932a4e2aa6e3fc3ec73b971f3f80143)
> * [Add edit URL when request fails test case](https://github.com/WordPress/gutenberg/pull/36099/commits/aa754c9c759fd4269c671187e089dc0e7654624c)

Thanks! The tests are looking good 👍🏾

fluiddot commented on packages/components/src/mobile/link-settings/index.native.js:
```diff
@@ -174,6 +175,13 @@ function LinkSettings( {
 	}, [ urlInputValue, labelInputValue, linkRelInputValue, setAttributes ] );
 
 	const onCloseSettingsSheet = useCallback( () => {
+		if ( onCloseSettingsSheetConsumed.current ) {
+			onCloseSettingsSheetConsumed.current = false;
+			return;
+		}
```
I think this approach might prevent the logic of `onCloseSettingsSheet` to be executed if the handler is only called once, which is the common behavior. I was thinking about a different approach that resets this value when the `isVisible` props changes from `false` to `true`:

```diff
diff --git a/packages/components/src/mobile/link-settings/index.native.js b/packages/components/src/mobile/link-settings/index.native.js
index f373f54450317e532758d5d277e155bc88e94001..0abebbdeb3fad62cc8a873ff081c97242cafdabe 100644
--- a/packages/components/src/mobile/link-settings/index.native.js
+++ b/packages/components/src/mobile/link-settings/index.native.js
@@ -124,6 +124,10 @@ function LinkSettings( {
 
 	useEffect( () => {
 		const isSettingSheetOpen = isVisible || editorSidebarOpened;
+		if ( isSettingSheetOpen ) {
+			onCloseSettingsSheetConsumed.current = false;
+		}
+
 		if ( options.url.autoFill && isSettingSheetOpen && ! url ) {
 			getURLFromClipboard();
 		}
@@ -176,7 +180,6 @@ function LinkSettings( {
 
 	const onCloseSettingsSheet = useCallback( () => {
 		if ( onCloseSettingsSheetConsumed.current ) {
-			onCloseSettingsSheetConsumed.current = false;
 			return;
 		}
 

```

@jd-alexander  I tested the above change in different cases and so far looks like it's working nice, let me know if you could take a look and if you agree, I'll push the changes.

jd-alexander replied:
The suggestion here works well @fluiddot I tested it and pushed the changes. 💯

fluiddot replied:
Thank you very much @jd-alexander for testing and pushing the changes 🥇 , I'll do a final review and merge it 🎊 .

fluiddot commented:
> > I run all test cases and everything is working nice 🎊 , I only found a minor issue related to displaying the invalid URL error notice twice
> 
> Thanks for spotting this error @fluiddot I resolved this issue by storing if the bottom sheet's close event was consumed in an instance variable via `useRef`. [3255826](https://github.com/WordPress/gutenberg/commit/32558261ca127553258b3df7b22983d3f331654f)
> 
> https://github.com/WordPress/gutenberg/blob/32558261ca127553258b3df7b22983d3f331654f/packages/components/src/mobile/link-settings/index.native.js#L96
> 
> The reason this occurs is because `onClose` is triggered in these two locations within the `LinkSettings` component.
> 
> 1. https://github.com/WordPress/gutenberg/blob/32558261ca127553258b3df7b22983d3f331654f/packages/components/src/mobile/link-settings/index.native.js#L102
> 2. https://github.com/WordPress/gutenberg/blob/ab7e8aa0312b2e5264af57c771a04cd9702370d3/packages/components/src/mobile/link-settings/index.native.js#L239
> 
> Let me know if this approach is optimal. Thank you 🙇🏾

@jd-alexander thank you very much for addressing the issue ❤️ ! I've just reviewed the changes and provided a change suggestion in [this comment](https://github.com/WordPress/gutenberg/pull/36099/files#r747487656).

jd-alexander committed:
onCloseSettingsSheetConsumed is set to false when isVisible is true.

jd-alexander committed:
Merge branch 'trunk' into rnmobile/embed-block-link-to

jd-alexander committed:
Added CHANGELOG entry to the unreleased section.

fluiddot commented:
I noticed that the CI check `End-to-End Tests / Admin - 1 (pull_request)` is failing with error:
```
FAIL packages/e2e-tests/specs/editor/blocks/image.test.js (71.564 s)
  ● Image › should replace, reset size, and keep selection

    expect(received).toBe(expected) // Object.is equality

    Expected: ""
    Received: "<!-- wp:image {\"id\":103,\"sizeSlug\":\"full\",\"linkDestination\":\"none\"} -->
    <figure class=\"wp-block-image size-full\"><img src=\"http://localhost:8889/wp-content/uploads/2021/11/ede1ce48-4ef8-4165-a74c-927833133aff.png\" alt=\"\" class=\"wp-image-103\"/></figure>
    <!-- /wp:image -->"

      108 | 		await page.keyboard.press( 'Backspace' );
      109 |
    > 110 | 		expect( await getEditedPostContent() ).toBe( '' );
          | 		                                       ^
      111 | 	} );
      112 |
      113 | 	it.skip( 'should place caret at end of caption after merging empty paragraph', async () => {

      at Object.<anonymous> (specs/editor/blocks/image.test.js:110:42)
          at runMicrotasks (<anonymous>)
```

However, this test is also failing in the commits in `trunk` branch ([reference](https://github.com/WordPress/gutenberg/runs/4186676637?check_suite_focus=true#step:6:9)), so looks safe to merge this PR with that failure as it's not related to the changes of this PR.

fluiddot reviewed (approved):
LGTM 🎊 ! Awesome work @jd-alexander 🥇 !

I've done the following test cases and verified that they passed 🟢 :
- Add Embed block - bottom sheet functionality remains ✅
- Edit Embed block - change the link in Block Settings ✅
- Edit Embed block - Invalid format link does not lead to validation error ✅ 
- Edit Embed block - Invalid link ✅ 

_Tested in iPhone 12 Pro Max simulator (iOS 15.0) and Samsung Galaxy S20 FE 5G (Android 11)._

fluiddot commented:
> However, this test is also failing in the commits in `trunk` branch ([reference](https://github.com/WordPress/gutenberg/runs/4186676637?check_suite_focus=true#step:6:9)), so looks safe to merge this PR with that failure as it's not related to the changes of this PR.

The `End-to-End Tests / Admin - 1 (pull_request` CI check is required so until it's fixed we can't merge the PR.

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-link-to

jd-alexander commented:
> LGTM 🎊 ! Awesome work @jd-alexander 🥇 !
> 
> I've done the following test cases and verified that they passed 🟢 :
> 
> * Add Embed block - bottom sheet functionality remains ✅
> * Edit Embed block - change the link in Block Settings ✅
> * Edit Embed block - Invalid format link does not lead to validation error ✅
> * Edit Embed block - Invalid link ✅
> 
> _Tested in iPhone 12 Pro Max simulator (iOS 15.0) and Samsung Galaxy S20 FE 5G (Android 11)._

Thanks for the review and tests @fluiddot 🙇🏾
